core/kotlinx-coroutines-core/src/internal/ExceptionsConstuctor.kt

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+import java.util.*
+import java.util.concurrent.locks.*
+import kotlin.concurrent.*
+
+private val cacheLock = ReentrantReadWriteLock()
+private val exceptionConstructors: WeakHashMap<Class<*>, (Throwable) -> Throwable?> = WeakHashMap()
```

Comment:
@cy6erGn0m had some concerns about this approach, though I don't see how `WeakReference<Class>` can prevent class unloading here.